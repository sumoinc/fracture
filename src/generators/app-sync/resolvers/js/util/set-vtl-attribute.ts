import { ResourceAttributeGenerator } from "../../../../../services/resource-attribute";
import { StructureAttribute } from "../../../../../services/structure-attribute";
import { TypescriptStrategy } from "../../../../ts/strategy";

export const setVtlAttribute = (attribute: StructureAttribute) => {
  const strategy = TypescriptStrategy.of(attribute.project);
  const generator = attribute.generator;
  const name = strategy.formatAttributeName(attribute.name);
  const shortName = strategy.formatAttributeName(attribute.shortName);

  // non-generated
  if (generator === ResourceAttributeGenerator.NONE) {
    return `"${shortName}": $util.dynamodb.toDynamoDBJson($ctx.args.input.${name}),`;
    // generated values
  } else {
    switch (generator) {
      case ResourceAttributeGenerator.CURRENT_DATE_TIME_STAMP:
        return `"${shortName}": "$util.time.nowISO8601()"`;
      case ResourceAttributeGenerator.GUID:
        return `"${shortName}": "$util.autoId()"`;
      case ResourceAttributeGenerator.TYPE:
        return `"${shortName}": "unknown"`;

      default:
        throw new Error(`Unknown generator: ${generator}`);
    }
  }
};
