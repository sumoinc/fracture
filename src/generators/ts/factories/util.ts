import { ValueOf } from "type-fest";
import { ResourceAttributeType } from "../../../services";

export const JsType = {
  STRING: "string",
  NUMBER: "number",
  BOOLEAN: "boolean",
  ARRAY: "array",
  MAP: "map",
  OTHER: "other",
} as const;

export const jsType = (
  resourceAttributeType: ValueOf<typeof ResourceAttributeType> | string
) => {
  switch (resourceAttributeType) {
    case ResourceAttributeType.GUID:
    case ResourceAttributeType.STRING:
    case ResourceAttributeType.EMAIL:
    case ResourceAttributeType.PHONE:
    case ResourceAttributeType.URL:
    case ResourceAttributeType.DATE:
    case ResourceAttributeType.TIME:
    case ResourceAttributeType.DATE_TIME:
    case ResourceAttributeType.JSON:
    case ResourceAttributeType.IPADDRESS:
      return JsType.STRING;
    case ResourceAttributeType.INT:
    case ResourceAttributeType.FLOAT:
    case ResourceAttributeType.TIMESTAMP:
    case ResourceAttributeType.COUNT:
    case ResourceAttributeType.AVERAGE:
    case ResourceAttributeType.SUM:
      return JsType.NUMBER;
    case ResourceAttributeType.BOOLEAN:
      return JsType.BOOLEAN;
    case ResourceAttributeType.ARRAY:
      return JsType.ARRAY;
    case ResourceAttributeType.MAP:
      return JsType.MAP;
    default:
      return JsType.OTHER;
  }
};
