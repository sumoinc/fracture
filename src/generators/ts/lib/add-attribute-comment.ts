import { Attribute } from "../../../model";
import { TypeScriptSource } from "../typescript-source";

export const addAttributeComment = (f: TypeScriptSource, a: Attribute) => {
  f.line(`/**`);
  a.comment.forEach((c) => f.line(` * ${c}`));
  f.line(` */`);
};
