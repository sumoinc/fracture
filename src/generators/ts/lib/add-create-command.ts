import { formatStringByNamingStrategy } from "../../../core/naming-strategy";
import { Shape } from "../../../model";
import { TypeScriptSource } from "../typescript-source";

export const addCreateCommand = (e: Shape) => {
  const command = `${e.fracture.namingStrategy.operations.commands.commandPrefix}-${e.fracture.namingStrategy.operations.crud.createName}-${e.name}-${e.fracture.namingStrategy.operations.commands.commandSuffix}`;
  const commandInput = `${e.fracture.namingStrategy.operations.commands.inputPrefix}-${e.fracture.namingStrategy.operations.crud.createName}-${e.name}-${e.fracture.namingStrategy.operations.commands.inputSuffix}`;
  const commandOutput = `${e.fracture.namingStrategy.operations.commands.outputPrefix}-${e.fracture.namingStrategy.operations.crud.createName}-${e.name}-${e.fracture.namingStrategy.operations.commands.inputSuffix}`;

  const fileName = formatStringByNamingStrategy(
    command,
    e.fracture.namingStrategy.ts.file
  );
  const f = new TypeScriptSource(e.service, `/commands/${fileName}.ts`);

  const commandInputName = formatStringByNamingStrategy(
    commandInput,
    e.fracture.namingStrategy.model.shapeName
  );

  const commandOutputName = formatStringByNamingStrategy(
    commandOutput,
    e.fracture.namingStrategy.model.shapeName
  );

  f.open(`export interface ${commandInputName} {`);
  f.line("foo: string;");
  f.close(`}`);
  f.line("\n");

  f.open(`export interface ${commandOutputName} {`);
  f.line("foo: string;");
  f.close(`}`);
  f.line("\n");

  f.open(`export const handler = async (event: any) => {`);
  f.line("return true;");
  f.close(`};`);
  f.line("\n");
};
