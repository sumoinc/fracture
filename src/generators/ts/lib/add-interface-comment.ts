import { Shape } from "../../../model";
import { TypeScriptSource } from "../typescript-source";

export const addInterfaceComment = (f: TypeScriptSource, e: Shape) => {
  f.line(`/**`);
  e.comment.forEach((c) => f.line(` * ${c}`));
  f.line(` */`);
};
