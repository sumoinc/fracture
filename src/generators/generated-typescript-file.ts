import { NodeProject } from "projen/lib/javascript";
import { GeneratedFile, GeneratedFileOptions } from "./generated-file";

export type GeneratedTypescriptFileOptions = GeneratedFileOptions;

export class GeneratedTypescriptFile extends GeneratedFile {
  constructor(
    public readonly project: NodeProject,
    filePath: string,
    options: GeneratedTypescriptFileOptions = {}
  ) {
    super(project, filePath, options);
  }

  preSynthesize() {
    this.addLine("// " + this.marker);
    this.addLine("");
    super.preSynthesize();
  }
}
