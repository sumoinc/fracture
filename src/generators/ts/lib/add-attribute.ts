import { addAttributeComment } from "./add-attribute-comment";
import { Attribute } from "../../../model";
import { TypeScriptSource } from "../typescript-source";

export const addAttribute = (f: TypeScriptSource, a: Attribute) => {
  addAttributeComment(f, a);

  const q = a.isRequired ? "" : "?";
  f.line(`${a.name}${q}: ${a.typeScriptType};`);
};
