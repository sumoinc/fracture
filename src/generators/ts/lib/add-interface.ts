import { addAttribute } from "./add-attribute";
import { addInterfaceComment } from "./add-interface-comment";
import { Entity } from "../../../model";
import { TypeScriptSource } from "../typescript-source";

export const addInterface = (f: TypeScriptSource, e: Entity) => {
  addInterfaceComment(f, e);

  f.open(`export interface ${e.name} {`);
  e.attributes.forEach((a) => {
    addAttribute(f, a);
  });
  f.close(`}`);
  f.line("\n");
};
