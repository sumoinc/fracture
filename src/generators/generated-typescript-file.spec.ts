import { NodeProject } from "projen/lib/javascript";
import { GeneratedTypescriptFile } from "./generated-typescript-file";
import { synthFile } from "../util/test-util";

describe("success conditions", () => {
  test("Smoke test", () => {
    const project = new NodeProject({
      name: "my-project",
      defaultReleaseBranch: "main",
    });

    new GeneratedTypescriptFile(project, "foo.ts").addLine("Hello, World!");

    const content = synthFile(project, "src/generated/foo.ts");
    expect(content).toBeTruthy();
    expect(content).toMatchSnapshot();
    //console.log(content);
  });
});
