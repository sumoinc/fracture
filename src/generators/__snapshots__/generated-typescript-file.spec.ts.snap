// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`success conditions Smoke test 1`] = `"Hello, World!"`;
